modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/VolatileTxStateMetaStorage.java

```diff
@@ -155,14 +187,62 @@ public void vacuum(long vacuumObservationTimestamp, long txnResourceTtl) {
             });
         });
 
-        LOG.info("Vacuum finished [vacuumObservationTimestamp={}, txnResourceTtl={}, vacuumizedTxnsCount={},"
-                + " markedAsInitiallyDetectedTxnsCount={}, alreadyMarkedTxnsCount={}, skippedFotFurtherProcessingUnfinishedTxnsCount={}].",
-                vacuumObservationTimestamp,
-                txnResourceTtl,
-                vacuumizedTxnsCount,
-                markedAsInitiallyDetectedTxnsCount,
-                alreadyMarkedTxnsCount,
-                skippedFotFurtherProcessingUnfinishedTxnsCount
+        beforeVacuum.apply(txIds)
+                .thenAccept(tuple -> {
+                    Set<UUID> successful = tuple.get1();
+
+                    for (UUID txId : successful) {
+                        txStateMap.compute(txId, (k, v) -> {
+                            if (v == null) {
+                                return null;
+                            } else {
+                                Long cleanupCompletionTs = timestamps.get(txId);
+
+                                return (cleanupCompletionTs != null && Objects.equals(cleanupCompletionTs, v.cleanupCompletionTimestamp()))
```

Comment:
Seems that cleanupCompletionTs != null is not needed here, we just may do v.cleanupCompletitionTimestamp.eqauls(cleanupCompletionTs), am I right?

Reply:
changed this condition